src/librustc/mir/interpret/allocation.rs

```diff
@@ -35,6 +36,21 @@ pub struct Allocation<Tag=(),Extra=()> {
     pub extra: Extra,
 }
 
+#[derive(Clone, Debug, Eq, PartialEq, PartialOrd, Ord, Hash, RustcEncodable, RustcDecodable)]
+pub struct AllocationBytes {
+    /// The actual bytes of the allocation.
+    /// Note that the bytes of a pointer represent the offset of the pointer
+    pub bytes: Vec<u8>,
+    /// Denotes undefined memory. Reading from undefined memory is forbidden in miri
+    pub undef_mask: UndefMask,
+}
+
+pub enum BytesOrElidedUndef<'a> {
+    /// Some bytes that may be undef but were not elided.
+    MaybeBytes(&'a [u8]),
+    /// An undef that has never been written to and was not yet allocated.
+    ElidedUndef,
+}
```

Comment:
I am not convinced that this is better than using `Option`. But if we stick with it, I don't think the "not yet allocated" part should matter, and the 2nd variant should be just `Undef`. There is no reason to leak the details of *when* we can easily know that the entire range is `Undef`.

Reply:
This is a proper `enum` because the `Some` variant may be easily misunderstood as meaning initialized bytes when in reality the range of `MaybeBytes` may still contain bytes in a logically uninitialized state.

Reply:
If you call `get_bytes` though, they will be initialized. And if you call `get_bytes_with_undef_and_ptr` or whatever we called it, you better read the docs.

Reply:
It will have to be its own type if the compiler were to adopt the optimization for constant values such as 'zeroed' as well. I'm not sure if that is a very good argument in support or if this should prepare such a future change.

Reply:
Not sure either. I am fine either way, but I really don't like the name `ElidedUndef`. ;)

Reply:
That's fine. Since the wrong interpretation is limited to the other variant, simply `Undef` or `Uninit` (better aligned with renaming the mask to `InitMask` and member ot `init_mask`) would also work.

Reply:
we may want to have an "all bytes zero" variant in the future, so the enum is in preparation for that. We could also just use `Option` everywhere and do the enum change when we actually try to implement the zeroes optimization

Reply:
Yes, I said I am okay with the enum, just not with the variant names. :)

Reply:
But @HeroicKatora already changed the name so this is resolved.